raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -584,6 +603,10 @@ private void becomeFollower(int epoch, OptionalInt leaderId) throws IOException
 
     private boolean maybeHandleError(RaftResponse.Inbound response, long currentTimeMs) throws IOException {
         ConnectionCache.ConnectionState connection = connections.getOrCreate(response.sourceId());
+        if (response.sourceId() == quorum.localId) {
```

Comment:
Here is a thing that I couldn't easily fix: in the event simulation test, we hit an issue when the target bootstrap server was randomly chosen the same as the current server, causing the connection hanging forever until expiration. We could discuss further on how to fix this problem, and more in general to avoid self-connection.